### DISCUSSION TITLE
搭建数据结构方面的问题

### DISCUSSION THREAD

**Discussion by ningninger** [liusy58/NJU_AdvancedProgramming_SP22#49] [Q&A] [answered]
当我使用c++的面向对象来设计信息管理系统的时候，我定义了一个类叫goods，并将goods的所有数据（包括id，name）放入它的private时，发现我文件的读入和输出需要把它列为友元。之后我定义了一个类数组，这样每次在我使用函数去修改这个类数组时，我必须都要把我的函数变成友元函数，那么我的函数都会变成友元函数，可是不是说尽量不要使用友元函数，因为它破坏了数据的封装么
还是说我只能去使用struct来去设计数据类型

**Comment by RicoloveFeng** [comment, marked as answer]
你可以用get/set函数来读写你的goods，再把中间数据拿去做文件读写。

```
// Good someGood{...};
int id = someGood.getId();
// do something with ''id''
someGood.setId(newId);
```

**Reply by ningninger** [reply]
那我定义一个类数组的思路是通的吗？之前c的过程设计时直接定义结构数组的话是可以直接使用外部函数去修改结构体，面向对象的话就要设置成友元了，感觉虽然更安全了，但限制更多了。

**Reply by binbin00** [reply]
你的定义一个类数组的思路是通的，更建议用STL。不一定要设置成友元，参考本楼的回答及例子。
